download/upgradeRecipe/upgradeRecipe6.2.adoc

```diff
@@ -0,0 +1,288 @@
+
+Before in `*.java`:
+[source, java]
+----
+public void addConstraintMatch(RuleContext kcontext, final int weight) {
+    ...
+    registerIntConstraintMatch(kcontext, 0, weight, new Runnable() {
+        public void run() {
+            ...
+        }
+    });
+}
+----
+
+After in `*.java`:
+[source, java]
+----
+public void addConstraintMatch(RuleContext kcontext, final int weight) {
+    ...
+    registerIntConstraintMatch(kcontext, 0, weight, new IntConstraintUndoListener() {
+        public void undo() {
+            ...
+        }
+    });
+}
+----
+
+====
+
+=== icon:exclamation-circle[] Major
+
+====
+If you have a custom `Move` implementation, now extract `AbstractMove`
+
+Before in `*.java`:
+[source, java]
+----
+public class CloudComputerChangeMove implements Move {...}
+----
+
+After in `*.java`:
+[source, java]
+----
+public class CloudComputerChangeMove extends AbstractMove {...}
+----
+
+====
+
+=== icon:info-circle[] Minor
+
+[.hidden-section]
+====
+The interface `Move` has a new method `getSimpleMoveTypeDescription()`.
+Extend `AbstractMove` so to avoid having to implement it.
+
+====
+
+=== icon:info-circle[] Minor
+
+[.hidden-section]
+====
+If you have a `@ValueRangeProvider` on an entity class (instead of the `Solution` class),
+then it's now safe to use the `<pillarChangeMoveSelector>` and `<pillarSwapMoveSelector>` as is.
+It's no longer needed to filter out swaps or changes
+which could put a value in an entity's variable that's not in its value range.
+
+Before in `*SolverConfig.xml` and `*BenchmarkConfig.xml`:
+[source, xml]
+----
+  <pillarChangeMoveSelector>
+    <filterClass>...ValidChangesOnlyPillarChangeMoveFilter</filterClass>
+  </pillarChangeMoveSelector>
+  <pillarSwapMoveSelector>
+    <filterClass>...ValidSwapsOnlyPillarSwapMoveFilter</filterClass>
+  </pillarSwapMoveSelector>
+----
+
+After in `*SolverConfig.xml` and `*BenchmarkConfig.xml`:
+[source, xml]
+----
+  <pillarChangeMoveSelector/>
+  <pillarSwapMoveSelector/>
+----
+
+====
+
+=== icon:info-circle[] Minor
+
+[.hidden-section]
+====
+`ConstraintMatchTotal` are now naturally compared by `scoreLevel`, then `constraintPackage`, then `constraintName`
+(instead of by `constraintPackage`, then `constraintName`, then `scoreLevel`) for readability.
+
+====
+
+=== icon:info-circle[] Minor
+
+[.hidden-section]
+====
+The optional method `IncrementalScoreCalculator.buildScoreCorruptionAnalysis(IncrementalScoreCalculator)` has been removed.
+Instead, to get a pretty score corruption analysis, implement the new interface `ConstraintMatchAwareIncrementalScoreCalculator`.
+This also enable your GUI to explain the score with an `IncrementalScoreCalculator`.
+See the example code in `MachineReassignmentIncrementalScoreCalculator.getConstraintMatchTotals()`.
+
+
+====
+
+== From 6.2.0.Beta1 to 6.2.0.Beta2
+
+
+=== icon:info-circle[] Minor
+
+[.hidden-section]
```

Comment:
I am also concerned about vertical space: the longer the list seems, the less likely users want to upgrade. So those h3 titles that just say "MINOR" are a bit a waste of space, it would be nicer if they describe the actual issue (and yes historically, those first sentences aren't short enough, but in the future we can try to stick to 80-140 chars max).

Here's how I 'd imagine a desired way (I realize this is probably not possible):

    [.minor]
    === The element `<deciderScoreComparatorFactory>` (which was deprecated, not documented, broken and clearly marked as not backwards compatible) has been removed.
    Instead, use strategic oscillation.
    Before in ...

Then it would automatically prefix the icon, the word "MINOR" and because it's minor and not major, automatically the the section hiding.